### PR TITLE
Using Virtuoso to Support Different KGs

### PR DESCRIPTION
**Pull request by HaishuoFang**
# What does this PR do?

Add a virtuoso image for Freebase and other KGs

### CODE REVIEW

File: docker-compose.ytt.yaml [hunk 1]
```diff
@@ -135,6 +135,29 @@ services:
     volumes:
       - db-data:/var/lib/postgresql/data
       - ./postgres/init:/docker-entrypoint-initdb.d/
+  
+  virtuoso:
+    image: tenforce/virtuoso:latest
```
Comment (timbmg) [inline docker-compose.ytt.yaml:140]: Ideally, do not se the latest tag but a specific version. This way we can make sure that the things we rely on do not change.

File: docker-compose.ytt.yaml [hunk 2]
```diff
@@ -135,6 +135,29 @@ services:
...
+      VIRT_$Database_Striping: 0
+      VIRT_$Database_TempStorage: TempDatabase
+      VIRT_$Parameters_O_DIRECT: 1
+      VIRT_$Parameters_ServerThreads: 100
+      VIRT_$HTTPServer_ServerThreads: 15
+      VIRT_$SPARQL_MaxQueryCostEstimationTime: 600
+      VIRT_$SPARQL_MaxQueryExecutionTime: 180
+
+    volumes:
+      - ./virtuoso_db:/data/virtuoso_db
```
Comment (timbmg) [inline docker-compose.ytt.yaml:158]: Use a volume managed by docker compose. In the volumes section add:
```yaml
virtuoso:
  # ...
  volumes:
      - virtuoso-db:/data/virtuoso_db
volumes:
  # ...
  virtuoso-db:
    driver: local

File: docker-compose.ytt.yaml [hunk 3]
```diff
@@ -135,6 +135,29 @@ services:
...
+    ports:
```
Comment (timbmg) [inline docker-compose.ytt.yaml:159]: Does it need to be exposed to the host? I suppose the skills, datastores and other services use the docker network to communicate with it.

File: docker-compose.ytt.yaml (hunk 1 above)
Comment (timbmg) [reply docker-compose.ytt.yaml:140]: [virtuoso7.2.5](https://hub.docker.com/layers/tenforce/virtuoso/virtuoso7.2.5/images/sha256-de97286328aa0babb9e06e9626321d91363ba3f260529b9083d1fa02f36ad664?context=explore) seems to be the latest tag.

File: docker-compose.ytt.yaml [hunk 4]
```diff
@@ -135,6 +135,29 @@ services:
     volumes:
       - db-data:/var/lib/postgresql/data
       - ./postgres/init:/docker-entrypoint-initdb.d/
+  
+  virtuoso:
+    image: tenforce/virtuoso:latest
+    environment: 
+      SPARQL_UPDATE: "true"
+      VIRT_$Database_DatabaseFile: /data/virtuoso_db/virtuoso.db
```
Comment (timbmg) [inline docker-compose.ytt.yaml:143]: Are you sure that the env variables are supposed to contain `$`? 🤔 
From my understanding they should not:
> All properties defined in virtuoso.ini can be configured via the environment variables. The environment variable should be prefixed with VIRT_ and have a format like VIRT_$SECTION_$KEY. $SECTION and $KEY are case sensitive. They should be CamelCased as in virtuoso.ini. E.g. property ErrorLogFile in the Database section should be configured as VIRT_Database_ErrorLogFile=error.log.

https://hub.docker.com/r/tenforce/virtuoso

**Review by timbmg — CHANGES_REQUESTED** [review]

File: docker-compose.ytt.yaml (hunk 4 above)
Comment (HaishuoFang) [reply docker-compose.ytt.yaml:143]: Thanks! I changed it

File: docker-compose.ytt.yaml (hunk 3 above)
Comment (HaishuoFang) [reply docker-compose.ytt.yaml:159]: Yes. I also need to query it from external networks to do some research work.

File: docker-compose.ytt.yaml [hunk 5]
```diff
@@ -135,6 +135,31 @@ services:
...
+      VIRT_Database_TempStorage: TempDatabase
+      VIRT_TempDatabase_DatabaseFile: /data/virtuoso-temp.db
+      VIRT_TempDatabase_TransactionFile: /data/virtuoso-temp.trx
+      VIRT_Parameters_O_DIRECT: 1
+      VIRT_Parameters_ServerThreads: 100
+      VIRT_HTTPServer_ServerThreads: 15
+      VIRT_SPARQL_MaxQueryCostEstimationTime: 600
+      VIRT_SPARQL_MaxQueryExecutionTime: 180
+    volumes:
+      - ./virtuoso_db:/data/
```
Comment (timbmg) [inline docker-compose.ytt.yaml:160]: it should be without the `./` and with a `-` ie `virtuoso-db`. It needs to match the name in the volumes.

**Pull request merged**